### PR TITLE
wal: do not call fsync when it is not necessary

### PR DESCRIPTION
**Pull request by xiang90**
We save commit index changes as a hint. So when a node restarts it can fast move forward commit index and then apply the entries.

But this is an optimization, and does not require fsync.

from raft paper:

```
Persistent state on all servers:
(Updated on stable storage before responding to RPCs)
- currentTerm 
- votedFor 
- log[]
```

### CODE REVIEW

File: wal/wal.go [hunk 1]
```diff
@@ -543,3 +548,10 @@ func (w *WAL) SaveSnapshot(e walpb.Snapshot) error {
 func (w *WAL) saveCrc(prevCrc uint32) error {
 	return w.encoder.encode(&walpb.Record{Type: crcType, Crc: prevCrc})
 }
+
+func mustSync(st, prevst raftpb.HardState, entsnum int) bool {
```
Comment (heyitsanthony) [inline wal/wal.go]: cite figure 2 from the paper as justification?

**Comment by heyitsanthony** [comment]
lgtm

**Comment by gyuho** [comment]
lgtm.

File: wal/wal.go (hunk 1 above)
Comment (xiang90) [inline wal/wal.go]: added.

**Pull request merged**